pkg/machinehandler/machinehandler.go

```diff
@@ -0,0 +1,174 @@
+package machinehandler
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	ErrUnstructuredFieldNotFound = fmt.Errorf("field not found")
+)
+
+type MachineHandler struct {
+	APIGroup string
+	Client   client.Client
+	Config   *rest.Config
+	Ctx      context.Context
+	machines []unstructured.Unstructured
+}
+
+// machineAddress contains information for the node's address.
+type machineAddress struct {
+	// Machine address type, one of Hostname, ExternalIP or InternalIP.
+	Type string `json:"type"`
+
+	// The machine address.
+	Address string `json:"address"`
+}
+
+// MachineAddresses is a slice of MachineAddress items to be used by infrastructure providers.
+type machineAddresses []machineAddress
+
+// ListMachines list all machines using given client
+func (m *MachineHandler) ListMachines() error {
+	APIVersion, err := m.getAPIGroupPreferredVersion()
+	if err != nil {
+		return err
+	}
+
+	unstructuredMachineList := &unstructured.UnstructuredList{}
+	unstructuredMachineList.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   m.APIGroup,
+		Kind:    "MachineList",
+		Version: APIVersion,
+	})
+	if err := m.Client.List(m.Ctx, unstructuredMachineList); err != nil {
+		return err
+	}
+
+	m.machines = unstructuredMachineList.Items
+
+	return nil
+}
```

Comment:
I wonder instead if we should create a "client.List" implementation that returns our own internal machine structure as a list

If we were to define a struct something like
```
type Machine struct {
  ObjectMeta metav1.ObjectMeta
  Status struct {
    NodeRef metav1.ObjectReference
    NodeAddresses corev1.NodeAddress
  }
}
```

And then used something like https://github.com/mitchellh/mapstructure, we could convert the unstructured data into our new type (preserving only the values that we care about) and the rest of the logic in the approver would be pretty simplistic and would just have to read from the fields as it does today

Reply:
that's a good idea, it can make the implementation cleaner.